README.md
<!-- ix-docs-ignore -->
![imgix logo](https://assets.imgix.net/sdk-imgix-logo.svg)

`imgix-python` is a client library for generating image URLs with [imgix](https://www.imgix.com/).

[![Version](https://img.shields.io/pypi/v/imgix.svg)](https://pypi.org/project/imgix/)
[![Build Status](https://travis-ci.com/imgix/imgix-python.svg?branch=main)](https://travis-ci.com/imgix/imgix-python)
![Downloads](https://img.shields.io/pypi/dm/imgix)
![Python Versions](https://img.shields.io/pypi/pyversions/imgix)
[![License](https://img.shields.io/github/license/imgix/imgix-python)](https://github.com/imgix/imgix-python/blob/main/LICENSE)
[![FOSSA Status](https://app.fossa.com/api/projects/git%2Bgithub.com%2Fimgix%2Fimgix-python.svg?type=shield)](https://app.fossa.com/projects/git%2Bgithub.com%2Fimgix%2Fimgix-python?ref=badge_shield)

---
<!-- /ix-docs-ignore -->

- [Installation](#installation)
- [Usage](#usage)
- [Signed URLs](#signed-urls)
- [Disabled Path Encoding](#disabled-path-encoding)
- [Srcset Generation](#srcset-generation)
    * [Fixed-Width Images](#fixed-width-images)
        + [Variable Quality](#variable-quality)
    * [Fluid-Width Images](#fluid-width-images)
        + [Custom Widths](#custom-widths)
        + [Width Ranges](#width-ranges)
        + [Width Tolerance](#width-tolerance)
        + [Explore Target Widths](#explore-target-widths)
    * [Usage with UTF-8](#usage-with-utf-8)
- [The `ixlib` Parameter](#the-ixlib-parameter)
- [Testing](#testing)
- [License](#license)

## Installation

``` bash
pip install imgix
```

## Usage

To begin creating imgix URLs, import the imgix library and create a URL builder. The URL builder can be reused to create URLs for any images on the domains it is provided.

``` python
>>> from imgix import UrlBuilder
>>> ub = UrlBuilder("demo.imgix.net")
>>> ub.create_url("bridge.png", {'w': 100, 'h': 100})
'https://demo.imgix.net/bridge.png?h=100&w=100'

```

_HTTPS_ support is enabled by default. _HTTP_ can be toggled on by setting `use_https` to `False`:

``` python
>>> from imgix import UrlBuilder
>>> ub = UrlBuilder("demo.imgix.net", use_https=False)
>>> ub.create_url("/bridge.png", {'w': 100, 'h': 100})
'http://demo.imgix.net/bridge.png?h=100&w=100'

```

## Signed URLs

To produce a signed URL, you must enable secure URLs on your source and then provide your signature key to the URL builder.

``` python
>>> from imgix import UrlBuilder
>>> ub = UrlBuilder("demo.imgix.net", sign_key="test1234")
>>> ub.create_url("/bridge.png", {'w': 100, 'h': 100})
'https://demo.imgix.net/bridge.png?h=100&w=100&s=bb8f3a2ab832e35997456823272103a4'

```

## Disabled Path Encoding

Path encoding is enabled by default. It can be toggled off by setting `disable_path_encoding` to `True` in the optional `options` paramater in `create_url()` and `create_srcset()` functions:

```python
>>> from imgix import UrlBuilder
>>> ub = UrlBuilder("sdk-test.imgix.net")
>>> ub.create_url(" <>[]{}|^%.jpg", params={'w': 100, 'h': 100}, options={'disable_path_encoding': True})
'https://sdk-test.imgix.net/ <>[]{}|^%.jpg?h=100&w=100'
```

Normally this would output a source URL like `https://demo.imgix.net/%20%3C%3E%5B%5D%7B%7D%7C%5E%25.jpg?h=100&2=100`, but since path encoding is disabled, it will output a source URL like `https://sdk-test.imgix.net/ <>[]{}|^%.jpg?h=100&w=100`.

```python
>>> from imgix import UrlBuilder
>>> ub = UrlBuilder("sdk-test.imgix.net")
>>> ub.create_srcset("image<>[]{} 123.png", widths=[100], options={'disable_path_encoding': True})
'https://sdk-test.imgix.net/image<>[]{} 123.png?w=100 100w'
```
Normally this would output a source URL like `https://sdk-test.imgix.net/image%3C%3E%5B%5D%7B%7D%20123.png?&w=100 100w`, but since path encoding is disabled, it will output a source URL like `https://sdk-test.imgix.net//image<>[]{} 123.png?w=100 100w`.

## Srcset Generation

The imgix-python package allows for generation of custom srcset attributes, which can be invoked through the `create_srcset` method. By default, the generated srcset will allow for responsive size switching by building a list of image-width mappings.

``` python
import os
from imgix import UrlBuilder

# Keep Your Secrets Safe!
SECRET = os.getenv("IX_SIGN_KEY")
ub = UrlBuilder("demos.imgix.net", sign_key=SECRET)
srcset = ub.create_srcset("image.png")
```

The above will produce the following srcset attribute value which can then be served to the client: 

``` html
https://demos.imgix.net/image.png?w=100&s=e415797545a77a9d2842dedcfe539c9a 100w,
https://demos.imgix.net/image.png?w=116&s=b2da46f5c23ef13d5da30f0a4545f33f 116w,
https://demos.imgix.net/image.png?w=135&s=b61422dead929f893c04b8ff839bb088 135w,
                                        ...
https://demos.imgix.net/image.png?w=7401&s=ad671301ed4663c3ce6e84cb646acb96 7401w,
https://demos.imgix.net/image.png?w=8192&s=a0fed46e2bbcc70ded13dc629aee5398 8192w
```

### Fixed-Width Images

In cases where enough information is provided about an image's dimensions, `create_srcset` will instead build a srcset that will allow for an image to be served at different resolutions. The parameters taken into consideration when determining if an image is fixed-width are `w`, `h`, and `ar`.

By invoking `create_srcset` with either a width **or** the height and aspect ratio (along with `fit=crop`, typically) provided, a different srcset will be generated for a fixed-width image instead.

``` python
from imgix import UrlBuilder
>>> ub = UrlBuilder("demos.imgix.net", sign_key="my-token")
>>> srcset = ub.create_srcset("image.png", {'h':800, 'ar':'3:2', 'fit':'crop'})

```

Will produce the following attribute value:

``` html
https://demos.imgix.net/image.png?ar=3%3A2&dpr=1&fit=crop&h=800&s=6cf5c443d1eb98bc3d96ea569fcef088 1x,
https://demos.imgix.net/image.png?ar=3%3A2&dpr=2&fit=crop&h=800&s=d60a61a5f34545922bd8dff4e53a0555 2x,
https://demos.imgix.net/image.png?ar=3%3A2&dpr=3&fit=crop&h=800&s=590f96aa426f8589eb7e449ebbeb66e7 3x,
https://demos.imgix.net/image.png?ar=3%3A2&dpr=4&fit=crop&h=800&s=c89c2fd3148957647e86cfc32ba20517 4x,
https://demos.imgix.net/image.png?ar=3%3A2&dpr=5&fit=crop&h=800&s=3d73af69d78d49eef0f81b4b5d718a2c 5x
```

By default, this library generates a `srcset` with pixel density values of `1` through `5`.
These target ratios can be controlled by using the `devicePixelRatios` parameters.

```python
from imgix import UrlBuilder
client = UrlBuilder("demo.imgix.net")
client.create_srcset(
  "image.jpg",
  { "w": 100 },
  {
    "device_pixel_ratios": [ 1, 2, 3 ]
  }
)
```
Will produce the following attribute value:
```html
https://demo.imgix.net/image.jpg?dpr=1&ixlib=python-3.2.1&q=75&w=100 1x,
https://demo.imgix.net/image.jpg?dpr=2&ixlib=python-3.2.1&q=50&w=100 2x,
https://demo.imgix.net/image.jpg?dpr=3&ixlib=python-3.2.1&q=35&w=100 3x'
```

For more information to better understand srcset, we highly recommend
[Eric Portis' "Srcset and sizes" article](https://ericportis.com/posts/2014/srcset-sizes/) which goes into depth about the subject.

#### Variable Quality

This library will automatically append a variable `q` parameter mapped to each `dpr` parameter when generating a [fixed-width image](#fixed-width-images) srcset. This technique is commonly used to compensate for the increased file size of high-DPR images.

Since high-DPR images are displayed at a higher pixel density on devices, image quality can be lowered to reduce overall file size––without sacrificing perceived visual quality. For more information and examples of this technique in action, see [this blog post](https://blog.imgix.com/2016/03/30/dpr-quality).

This behavior will respect any overriding `q` value passed in as a parameter. Additionally, it can be disabled altogether by passing `disable_variable_quality = true` to `create_srcset`.

This behavior specifically occurs when a [fixed-width image](#fixed-width-images) is rendered, for example:

```python
# Note that `params={"w": 100}` allows `create_srcset` to _infer_ the creation
# of a DPR based srcset attribute for fixed-width images.
ub = imgix.UrlBuilder('demo.imgix.net')
# Set `disable_variable_quality` to True to disable variable quality.
srcset = ub.create_srcset('image.jpg', params={"w": 100}, disable_variable_quality=False)
```

The above will generate a srcset with the following `q` to `dpr` query `params`:

```html
https://demo.imgix.net/image.jpg?w=100&dpr=1&q=75 1x,
https://demo.imgix.net/image.jpg?w=100&dpr=2&q=50 2x,
https://demo.imgix.net/image.jpg?w=100&dpr=3&q=35 3x,
https://demo.imgix.net/image.jpg?w=100&dpr=4&q=23 4x,
https://demo.imgix.net/image.jpg?w=100&dpr=5&q=20 5x
```

By default, this library will automatically append a variable `q` parameter mapped to each `dpr` parameter when generating a [fixed-width image](#fixed-width-images) srcset.

To customize variable qualities, you can pass a `variable_qualities` dictionary in the `options` while creating srcset as below:
```python
from imgix import UrlBuilder
client = UrlBuilder("demo.imgix.net")
client.create_srcset("image.jpg", {"w": 100}, {"variable_qualities": {1: 45, 2: 30, 3: 20, 4: 15, 5: 10}})
```
The above script will produce the following output:
```bash
https://demo.imgix.net/image.jpg?dpr=1&q=45&w=100 1x
https://demo.imgix.net/image.jpg?dpr=2&q=30&w=100 2x
https://demo.imgix.net/image.jpg?dpr=3&q=20&w=100 3x
https://demo.imgix.net/image.jpg?dpr=4&&q=15&w=100 4x
https://demo.imgix.net/image.jpg?dpr=5&q=10&w=100 5x
``` 

You can also pass `variable_qualities` along with the `device_pixel_ratios` option as below:
```python
from imgix import UrlBuilder
client = UrlBuilder("demo.imgix.net")
client.create_srcset(
  "image.jpg",
  { "w": 100 },
  {
    "device_pixel_ratios": [ 1, 2, 3 ],
    "variable_qualities": { 1: 45, 2: 30, 3: 20 }
  }
)
```
The above script will produce the following output:
```html
https://testing.imgix.net/image.jpg?dpr=1&q=45&w=100 1x,
https://testing.imgix.net/image.jpg?dpr=2&q=30&w=100 2x,
https://testing.imgix.net/image.jpg?dpr=3&q=20&w=100 3x
```

### Fluid-Width Images

#### Custom Widths

In situations where specific widths are desired when generating `srcset` pairs, a user can specify them by passing an array of positive integers as `widths`:

``` python
>>> from imgix import UrlBuilder
>>> builder = UrlBuilder('demo.imgix.net')
>>> builder.create_srcset('image.jpg', widths=[144, 240, 320, 446, 640])
'https://demo.imgix.net/image.jpg?w=144 144w,\nhttps://demo.imgix.net/image.jpg?w=240 240w,\nhttps://demo.imgix.net/image.jpg?w=320 320w,\nhttps://demo.imgix.net/image.jpg?w=446 446w,\nhttps://demo.imgix.net/image.jpg?w=640 640w'

```

```html
https://demo.imgix.net/image.jpg?w=144 144w,
https://demo.imgix.net/image.jpg?w=240 240w,
https://demo.imgix.net/image.jpg?w=320 320w,
https://demo.imgix.net/image.jpg?w=446 446w,
https://demo.imgix.net/image.jpg?w=640 640w
```

**Note**: in situations where a `srcset` is being rendered as a [fixed-width](#fixed-width-images) srcset, any custom `widths` passed in will be ignored.

Additionally, if both `widths` and a width `tol`erance are passed to the `create_srcset` method, the custom widths list will take precedence.

#### Width Ranges

In certain circumstances, you may want to limit the minimum or maximum value of the non-fixed `srcset` generated by the `create_srcset` method. To do this, you can specify the widths at which a srcset should `start` and `stop`:

```python
>>> from imgix import UrlBuilder
>>> ub = UrlBuilder('demo.imgix.net')
>>> ub.create_srcset('image.jpg', start=500, stop=2000)
'https://demo.imgix.net/image.jpg?w=500 500w,\nhttps://demo.imgix.net/image.jpg?w=580 580w,\nhttps://demo.imgix.net/image.jpg?w=673 673w,\nhttps://demo.imgix.net/image.jpg?w=780 780w,\nhttps://demo.imgix.net/image.jpg?w=905 905w,\nhttps://demo.imgix.net/image.jpg?w=1050 1050w,\nhttps://demo.imgix.net/image.jpg?w=1218 1218w,\nhttps://demo.imgix.net/image.jpg?w=1413 1413w,\nhttps://demo.imgix.net/image.jpg?w=1639 1639w,\nhttps://demo.imgix.net/image.jpg?w=1901 1901w,\nhttps://demo.imgix.net/image.jpg?w=2000 2000w'

```

Formatted version of the above srcset attribute:

``` html
https://demo.imgix.net/image.jpg?w=500 500w,
https://demo.imgix.net/image.jpg?w=580 580w,
https://demo.imgix.net/image.jpg?w=673 673w,
https://demo.imgix.net/image.jpg?w=780 780w,
https://demo.imgix.net/image.jpg?w=905 905w,
https://demo.imgix.net/image.jpg?w=1050 1050w,
https://demo.imgix.net/image.jpg?w=1218 1218w,
https://demo.imgix.net/image.jpg?w=1413 1413w,
https://demo.imgix.net/image.jpg?w=1639 1639w,
https://demo.imgix.net/image.jpg?w=1901 1901w,
https://demo.imgix.net/image.jpg?w=2000 2000w'
```

#### Width Tolerance

The `srcset` width `tol`erance dictates the maximum `tol`erated difference between an image's downloaded size and its rendered size.

For example, setting this value to `0.10` means that an image will not render more than 10% larger or smaller than its native size. In practice, the image URLs generated for a width-based srcset attribute will grow by twice this rate.

A lower tolerance means images will render closer to their native size (thereby increasing perceived image quality), but a large srcset list will be generated and consequently users may experience lower rates of cache-hit for pre-rendered images on your site.

By default, srcset width `tol`erance is set to 0.08 (8 percent), which we consider to be the ideal rate for maximizing cache hits without sacrificing visual quality. Users can specify their own width tolerance by providing a positive scalar value as width `tol`erance:

```python
>>> import imgix
>>> ub = imgix.UrlBuilder('demo.imgix.net')
>>> srcset = ub.create_srcset('image.jpg', start=100, stop=384, tol=0.20)

```

In this case, the `width_tolerance` is set to 20 percent, which will be reflected in the difference between subsequent widths in a srcset pair:

```html
https://demo.imgix.net/image.jpg?w=100 100w,
https://demo.imgix.net/image.jpg?w=140 140w,
https://demo.imgix.net/image.jpg?w=196 196w,
https://demo.imgix.net/image.jpg?w=274 274w,
https://demo.imgix.net/image.jpg?w=384 384w
```

#### Explore Target Widths

The `target_widths` function is used internally to generate lists of target widths to be used in calls to `create_srcset`.

It is a way to generate, play with, and explore different target widths separately from srcset attributes. One way of generating a srcset attribute is:

```python
srcset = ub.create_srcset('image.jpg', start=300, stop=3000, tol=0.13)
```

The above is convenient if `start`, `stop`, and `tol`erance are known in advance. Another approach is to use `target_widths` to determine which combination of values for `start`, `stop`, and `tol`erance work best.

```python
>>> from imgix import UrlBuilder, target_widths
>>> # Create
>>> widths = target_widths(300, 3000, 0.13)
>>> widths
[300, 378, 476, 600, 756, 953, 1200, 1513, 1906, 2401, 3000]
>>> # Explore
>>> sm, md, lg = widths[:3], widths[3:7], widths[7:]
>>> widths = [w for w in widths[1::2]]
>>> widths
[378, 600, 953, 1513, 2401]
>>> # Serve
>>> ub = UrlBuilder('demo.imgix.net')
>>> srcset = ub.create_srcset('image.png', widths=widths)
>>> srcset
'https://demo.imgix.net/image.png?w=378 378w,\nhttps://demo.imgix.net/image.png?w=600 600w,\nhttps://demo.imgix.net/image.png?w=953 953w,\nhttps://demo.imgix.net/image.png?w=1513 1513w,\nhttps://demo.imgix.net/image.png?w=2401 2401w'

```

### Usage with UTF-8

For usage with non-ASCII characters, please be sure that your project's source files specify UTF-8 encoding:

``` python
# -*- coding: utf-8 -*-
```

If you don't add this encoding, and you have an image with the name 'tiburón.jpeg', for example, you will get the following error trying to run your script:

``` python
SyntaxError: Non-ASCII character '***' in file test.py on line 6, but no encoding declared; see http://www.python.org/peps/pep-0263.html for details
```

## The `ixlib` Parameter

For security and diagnostic purposes, we sign all requests with the language and version of library used to generate the URL.

This can be disabled by setting `include_library_param` to `False` like so:

``` python
UrlBuilder('demo.imgix.net', include_library_param=False)
```

## Testing

Run the following to execute the project's tests and code linter:

``` bash
tox
```

If you have cloned this repo or downloaded it locally, you can also run `python -m doctest -v README.md` to test the examples in this readme.

## License
[![FOSSA Status](https://app.fossa.com/api/projects/git%2Bgithub.com%2Fimgix%2Fimgix-python.svg?type=large)](https://app.fossa.com/projects/git%2Bgithub.com%2Fimgix%2Fimgix-python?ref=badge_large)


CHANGELOG.md
# Changelog

## [4.0.0](https://github.com/imgix/imgix-python/compare/3.2.1...4.0.0) (2022-12-08)

### Breaking Changes

- feat: drop python 2 support ([#88](https://github.com/imgix/imgix-python/pull/88))
- fix: fixed-height produces dpr-based srcset ([#88](https://github.com/imgix/imgix-python/pull/88))
- feat: encode file path components ([#88](https://github.com/imgix/imgix-python/pull/88))

### Features

- feat: add path encoding to be disabled optionally ([#98](https://github.com/imgix/imgix-python/pull/98))
- feat: customize variable qualities in options ([#102](https://github.com/imgix/imgix-python/pull/102))
- feat: customize target device pixel ratios ([#103](https://github.com/imgix/imgix-python/pull/103))

## [3.2.1](https://github.com/imgix/imgix-python/compare/3.2.0...3.2.1) (2020-06-10)

- refactor: relax validation for min/max width values ([#80](https://github.com/imgix/imgix-python/pull/80))

## [3.2.0](https://github.com/imgix/imgix-python/compare/3.1.2...3.2.0) (2020-06-05)

- feat: create custom srcset ([#63](https://github.com/imgix/imgix-python/pull/63))
- feat: introduce variable image output quality ([#65](https://github.com/imgix/imgix-python/pull/65))
- fix: remove ensure even ([#72](https://github.com/imgix/imgix-python/pull/72))
- feat: throw exceptions from validators ([#77](https://github.com/imgix/imgix-python/pull/77))
- fix: convert tol to float ([#75](https://github.com/imgix/imgix-python/pull/75))

## [3.1.2](https://github.com/imgix/imgix-python/compare/3.1.1...3.1.2) (2020-03-11)

- Fix Python 2/3 compatibility issues
  ([\#57](https://github.com/imgix/imgix-python/pull/57))

## [3.1.1](https://github.com/imgix/imgix-python/compare/3.1.0...3.1.1) (2019-08-22)

- fix: include dpr parameter when generating fixed-width srcset
  ([\#50](https://github.com/imgix/imgix-python/pull/50))

## [3.1.0](https://github.com/imgix/imgix-python/compare/3.0.0...3.1.0) (2019-08-22)

- feat: add srcset generation
  ([\#48](https://github.com/imgix/imgix-python/pull/48))
- build(tox): improve code coverage reporting; parallelize testing
  ([\#49](https://github.com/imgix/imgix-python/pull/49))

## [3.0.0](https://github.com/imgix/imgix-python/compare/2.3.0...3.0.0) (2019-06-07)

- fix: remove deprecated domain sharding functionality
  ([\#44](https://github.com/imgix/imgix-python/pull/44))
- fix: remove deprecated [opts]{.title-ref} parameter
  ([\#46](https://github.com/imgix/imgix-python/pull/46))
- fix: remove deprecated [sign\_with\_library\_version]{.title-ref}
  parameter ([\#47](https://github.com/imgix/imgix-python/pull/47))

## [2.3.0](https://github.com/imgix/imgix-python/compare/2.2.0...2.3.0) (2019-06-06)

- feat: deprecate [domains]{.title-ref} in favor of
  [domain]{.title-ref}
  ([\#45](https://github.com/imgix/imgix-python/pull/45))

## [2.2.0](https://github.com/imgix/imgix-python/compare/2.1.0...2.2.0) (2019-05-07)

- deprecate domain sharding
  ([\#41](https://github.com/imgix/imgix-python/pull/41))([\#42](https://github.com/imgix/imgix-python/pull/42))

## 2.1.0 (2019-02-13)

- Domain validation added during [UrlBuilder]{.title-ref}
  initialization
- [sign\_with\_library\_version]{.title-ref} parameter from
  [UrlBuilder]{.title-ref} deprecated in favor of
  [include\_library\_param]{.title-ref}.

## 2.0.0 (2018-08-08)

- [UrlBuilder]{.title-ref}\'s [sign\_mode]{.title-ref} argument
  removed
- [opts]{.title-ref} parameter from
  [UrlBuilder.create\_url]{.title-ref} deprecated in favor of
  [params]{.title-ref}.

## 1.2.0 (2018-06-20)

- [sign\_mode]{.title-ref} argument deprecated
- License corrected to BSD-2-Clause.
- Docstrings added to classes and methods.

## 1.1.2 (2016-06-30)

- Proper encodeURIComponent-style URL encoding for web proxy sources.
  See \#21 for more information.

## 1.1.0 (2016-02-26)

- Added automatic Base64 encoding for all Base64 variant parameters.
- Properly encoding all query keys and values.

## 1.0.0 (2016-01-15)

- Change UrlBuilder\#create_url to accept dict instead of kwargs.
  This fixes an issue with reserved words that are also imgix params
  potentially causing errors.

## 0.2.1 (2016-01-15)

- Fixed a bug where any passed params that were falsy would not be
  passed through to imgix.

## 0.2.0 (2015-06-15)

- Introduces defaulting to HTTPS on all requests, per the
  imgix-blueprint.

## 0.1.0 (2015-06-11)

- Includes new functionality to sign every URL with an ixlib parameter
  for diagnostic purposes.

## 0.0.4 (2015-06-10)

- New README note about publishing packages


imgix/urlbuilder.py
# -*- coding: utf-8 -*-
import math
import re
import hashlib

from ._version import __version__


from base64 import urlsafe_b64encode
from urllib.parse import quote_plus, quote
from .validators import (
    validate_device_pixel_ratios,
    validate_min_max_tol,
    validate_variable_qualities,
    validate_widths,
)

from .constants import DPR_QUALITIES
from .constants import IMAGE_MAX_WIDTH as MAX_WIDTH
from .constants import IMAGE_MIN_WIDTH as MIN_WIDTH
from .constants import SRCSET_WIDTH_TOLERANCE as TOLERANCE
from .constants import SRCSET_DPR_TARGET_RATIOS as TARGET_RATIOS
from .constants import DOMAIN_PATTERN, SRCSET_TARGET_WIDTHS as TARGET_WIDTHS


class UrlBuilder(object):
    """
    Create imgix URLs

    The URL builder can be reused to create URLs for any images on the
    provided domain.

    Parameters
    ----------
    domain : str
        Domain to use while creating imgix URLs.
    use_https : bool
        If `True`, create HTTPS imgix image URLs. (default `True`)
    sign_key : str or None
        When provided, this key will be used to sign the generated image URLs.
        You can read more about URL signing on our docs:
        https://docs.imgix.com/setup/securing-images
    include_library_param : bool
        If `True`, each created URL is suffixed with 'ixlib' parameter
        indicating the library used for generating the URLs. (default `True`)

    Methods
    -------
    validate_domain(domain)
        Returns true if the supplied string parameter pattern matches a valid
        domain name accepted by imgix
    create_url(path, params=None)
        Create URL with the supplied path and `params` parameters dict.
    create_srcset(path, params=None)
        Create srcset attribute value with the supplied path and
        `params` parameters dict.
        Will generate a fixed-width DPR srcset if a width OR height and aspect
        ratio are passed in as parameters. Otherwise will generate a srcset
        with width-descriptor pairs.
    """

    def __init__(
        self, domain, use_https=True, sign_key=None, include_library_param=True
    ):

        self.validate_domain(domain)

        self._domain = domain
        self._sign_key = sign_key
        self._use_https = use_https
        self._include_library_param = include_library_param

    def validate_domain(self, domain):
        """
        Returns true if the supplied string parameter pattern matches a valid
        domain name accepted by imgix

        Parameters
        ----------
        domain : str

        Returns
        -------
        bool
        """

        err_str = str(
            "Domain must be passed in as fully-qualified domain names and "
            + "should not include a protocol or any path element, i.e. "
            + '"example.imgix.net".'
        )

        if re.match(DOMAIN_PATTERN, domain) is None:
            raise ValueError(err_str)

    def create_url(self, path="", params={}, options={}):
        """
        Create URL with supplied path, `params` parameters dict
        and optional `options` dict.

        Parameters
        ----------
        path : str
        params : dict
            Dictionary specifying URL parameters. Non-imgix parameters are
            added to the URL unprocessed. For a complete list of imgix
            supported parameters, visit https://docs.imgix.com/apis/url .
            (default None)
        options : dict
            Dictionary specifying URL options such as disabled_path_encoding.
            (default None)

        Returns
        -------
        str
            imgix URL
        """
        disable_path_encoding = options.get("disable_path_encoding", False)
        sanitized_path = self._sanitize_path(
            path, options={"disable_path_encoding": disable_path_encoding}
        )

        query_string = self._build_params(params)

        if self._sign_key:
            query_string = self._sign_url(sanitized_path, query_string)

        scheme = "https" if self._use_https else "http"

        return scheme + "://" + self._domain + sanitized_path + query_string

    def _sanitize_path(self, path, options={}):
        if not path:
            return ""

        _path = path[:]
        # If the path is prefixed with a forward slash,
        # remove it.
        if _path[0] == "/":
            _path = _path[1:]

        # Encode the path without a leading forward slash,
        # then add it back before returning.
        if options["disable_path_encoding"]:
            return "/" + _path
        elif _path.startswith("http"):
            return "/" + self._encode_proxy_path(_path)
        else:
            return "/" + self._encode_file_path(_path)

    def _encode_file_path(self, path):
        return quote(path, safe="/&$;=@,")

    def _encode_proxy_path(self, path):
        return quote_plus(path)

    def _build_params(self, params):
        # Stringify all param values.
        _params = {quote(k): str(v) for k, v in params.items()}

        if self._include_library_param:
            _params["ixlib"] = "python-" + __version__

        for k, v in _params.items():
            # Base64 encode any params whose keys end with '64'.
            # The encoding function uses '=' as padding so we replace
            # those with ''.
            if k.endswith("64"):
                # First we call encode on v to get a bytes-like object,
                # then after replacing any padding characters that may
                # be present, we call decode to get back a string object.
                _params[k] = (
                    urlsafe_b64encode(v.encode("utf-8"))
                    .replace(b"=", b"")
                    .decode("utf-8")
                )
            else:
                # quote_plus will encode SPACE (' ') as PLUS (+). If a
                # PLUS (+) is present, e.g. "Futura+Condensed Medium",
                # it will be encoded as "Futura%2BCondensed+Medium".
                _params[k] = quote_plus(v).replace("+", "%20")

        query_string = [f"{k}={_params[k]}" for k in sorted(_params.keys())]
        delimeter = "?" if query_string else ""
        return delimeter + "&".join(query_string)

    def _sign_url(self, prefixed_path, query_string):
        signature_base = self._sign_key + prefixed_path + query_string
        signature = hashlib.md5(signature_base.encode("utf-8")).hexdigest()
        delimeter = "&s=" if query_string else "?s="
        return query_string + delimeter + signature

    def create_srcset(self, path, params={}, options={}, **kwargs):
        """
        Create a srcset attribute.

        A srcset attribute consists of one or more non-empty URL. Each URL
        represents an image candidate string and each candidate string is
        separated by a comma (U+002C) character (,). Read more about the
        srcset attribute here:

        https://html.spec.whatwg.org/multipage/images.html#srcset-attributes

        This function produces two types of image candidate strings,

        * pixel density descriptors (x) and
        * width descriptors (w)

        Pixel density-described strings are produced when

        * a height (h) and an aspect (ar) ratio are present in `params`, or
        * only a width (w) is present in the `params`

        Example, a width (or a height _and_ aspect ratio):
        'https://example.test.com/image/path.png?dpr=1&w=320 1x'

        Width-described strings are produced if neither a width nor a
        height-aspect-ratio pair are present in `params`.

        Example, no width, no height, no aspect ratio:
        'https://example.test.com/image/path.png?w=100 100w'

        Parameters
        ----------
        path : str
            Path to the image file, e.g. 'image/path.png'.
        params : dict, optional
            Parameters that will be transformed into query parameters,
            including 'w' or 'ar' and 'h' if generating a pixel density
            described srcset, {} by default.
        options: dict, optional
            Options that will be used to generate the srcset,
            including 'disable_path_encoding', 'variable_qualities',
            'device pixel ratios', {} by default.
        start : int, optional
            Starting minimum width value, MIN_WIDTH by default.
        stop : int, optional
            Stopping maximum width value, MAX_WIDTH by default.
        tol : float, optional
            Tolerable amount of width value variation, TOLERANCE by default.
        widths: list, optional
            List of target widths, `TARGET_WIDTHS` by default.
        Returns
        -------
        str
            Srcset attribute string.
        """
        widths_list = kwargs.get("widths", None)
        if widths_list:
            validate_widths(widths_list)
            return self._build_srcset_pairs(
                path, params, options, targets=widths_list
            )

        # Attempt to assign `start`, `stop`, and `tol` from `kwargs`.
        # If the key does not exist, assign `None`.
        start = kwargs.get("start", None)
        stop = kwargs.get("stop", None)
        tol = kwargs.get("tol", None)

        # Attempt to generate the specified target widths.
        # Assign defaults where appropriate, then validate
        # this group. If validation succeeds, generate the
        # target widths.
        if start is None:
            start = MIN_WIDTH
        if stop is None:
            stop = MAX_WIDTH
        if tol is None:
            tol = TOLERANCE

        validate_min_max_tol(start, stop, tol)

        targets = target_widths(start=start, stop=stop, tol=tol)

        if "w" in params or "h" in params:
            disable_variable_quality = kwargs.get(
                "disable_variable_quality", False
            )
            return self._build_srcset_DPR(
                path,
                params,
                options,
                disable_variable_quality=disable_variable_quality,
            )
        else:
            return self._build_srcset_pairs(path, params, options, targets)

    def _build_srcset_pairs(
        self, path, params, options, targets=TARGET_WIDTHS
    ):
        # prevents mutating the params dict
        srcset_params = dict(params)
        srcset_entries = []

        for w in targets:
            srcset_params["w"] = w
            srcset_entries.append(
                self.create_url(path, srcset_params, options)
                + " "
                + str(w)
                + "w"
            )

        return ",\n".join(srcset_entries)

    def _build_srcset_DPR(
        self,
        path,
        params,
        options,
        targets=TARGET_RATIOS,
        disable_variable_quality=False,
    ):
        # If variable quality output is _not disabled_, then output
        # quality values, 'q', will vary in accordance with the
        # default `DPR_QUALITIES` [1x => q=75, ... 5x => 20].
        #
        # If 'q' is explicitly passed, it takes precedence over
        # the default `DPR_QUALITIES`. Right now, this means that
        # if 'q' is passed, it's value will be used for the output
        # quality of each dpr image (i.e. for 1x through 5x).
        #
        # Note: if `q` is passed with `params`, then it will be
        # present in the URL's query params whether or not
        # `disabled_variable_quality` is `True` or `False`.

        # prevents mutating the params dict
        srcset_params = dict(params)
        srcset_entries = []

        variable_qualities = options.get("variable_qualities", {})
        if variable_qualities:
            validate_variable_qualities(variable_qualities)
            targets = range(1, len(variable_qualities)+1)

        qualities = {**DPR_QUALITIES, **variable_qualities}

        device_pixel_ratios = options.get("device_pixel_ratios", [])
        if device_pixel_ratios:
            validate_device_pixel_ratios(device_pixel_ratios)
            targets = device_pixel_ratios

        for dpr in targets:
            srcset_params["dpr"] = dpr

            if not disable_variable_quality:
                quality = (
                    params.get("q", None)
                    or qualities.get("dpr")
                    or qualities.get(math.floor(dpr))
                )
                if quality:
                    srcset_params["q"] = quality

            srcset_entries.append(
                self.create_url(path, srcset_params, options)
                + " "
                + str(dpr)
                + "x"
            )

        return ",\n".join(srcset_entries)


def target_widths(start=MIN_WIDTH, stop=MAX_WIDTH, tol=TOLERANCE):
    """
    Generate a list of target widths.

    This function generates a list of target widths used to width-describe
    image candidate strings (URLs) within a srcset attribute.

    For example, if the target widths are [100, 200, 300], they would become:

    'https://example.test.com/image/path.png?w=100 100w
    https://example.test.com/image/path.png?w=200 200w
    https://example.test.com/image/path.png?w=300 300w'

    in the srcset attribute string. Read more about image candidate strings
    and width descriptors here:

    https://html.spec.whatwg.org/multipage/images.html#image-candidate-string

    Parameters
    ----------
    start : int, optional
        Starting minimum width value, MIN_WIDTH by default.
    stop : int, optional
        Stopping maximum width value, MAX_WIDTH by default.
    tol : float, optional
        Tolerable amount of image width-variation, TOLERANCE by default.

    Returns
    -------
    list
        A list of even integer values.
    """
    validate_min_max_tol(start, stop, tol)
    # If any value differs from the default, we're constructing a custom
    # target widths list.
    CUSTOM = any([tol != TOLERANCE, start != MIN_WIDTH, stop != MAX_WIDTH])

    if not CUSTOM:
        return TARGET_WIDTHS

    if start == stop:
        return [int(start)]

    resolutions = []

    while start < stop and start < MAX_WIDTH:
        resolutions.append(int(round(start)))
        start *= 1 + tol * 2

    # The most recently appended value may, or may not, be
    # the `stop` value. In order to be inclusive of the
    # stop value, check for this case and add it, if necessary.
    if resolutions[-1] < stop:
        resolutions.append(int(stop))

    return resolutions
